sympy/solvers/ode/nonhomogeneous.py

```diff
@@ -0,0 +1,512 @@
+
+2) _get_euler_characteristic_eq_sols(eq, func, match_obj) - It returns the solution of homogeneous
+    part of the linear euler ODE and the list of roots of characteristic equation
+
+3) _solve_variation_of_parameters(eq, func, roots, homogen_sol, order, match_obj, simplify_flag=True)-
+  Helper function for the method of variation of parameters and nonhomogeneous euler eq.
+
+4) _get_const_characteristic_eq_sols(r, func, order) - It returns the roots of characteristic equation
+    of constant coefficient linear ODE and list of collectterms which is later on used by simplification
+    to use collect on solution.
+
+5) _get_simplified_sol(sol, func, collectterms) - It is a helper function which collects the solution on
+    collectterms. Ideally this should be handled by odesimp.
+
+6) _undetermined_coefficients_match(expr, x, func=None, eq_homogeneous=S.Zero) - It returns a trial function
+    match if undetermined coefficients can be applied to ``expr``, and ``None`` otherwise.
+
+7) _solve_undetermined_coefficients(eq, func, order, match, trialset) - Helper function for the method of
+    undetermined coefficients.
+
+"""
+from collections import defaultdict
+from sympy.core import Add, S
+from sympy.core.function import  diff, expand, _mexpand, expand_mul
+from sympy.core.relational import Eq
+from sympy.core.symbol import Dummy, Wild
+from sympy.functions import exp, cos, cosh, im, log, re, sin, sinh, \
+    atan2, conjugate
+from sympy.integrals import Integral
+from sympy.polys import (Poly, RootOf, rootof, roots)
+from sympy.simplify import collect, simplify, separatevars, powsimp, trigsimp
+from sympy.utilities import numbered_symbols, default_sort_key
+from sympy.solvers.solvers import solve
+from sympy.matrices import wronskian
+from .subscheck import sub_func_doit
+from sympy.solvers.ode.ode import get_numbered_constants
+
+
+def _test_term(coeff, func, order):
+    r"""
+    Linear Euler ODEs have the form  K*x**order*diff(y(x), x, order) = F(x),
+    where K is independent of x and y(x), order>= 0.
+    So we need to check that for each term, coeff == K*x**order from
+    some K.  We have a few cases, since coeff may have several
+    different types.
+    """
+    x = func.args[0]
+    f = func.func
+    if order < 0:
+        raise ValueError("order should be greater than 0")
+    if coeff == 0:
+        return True
+    if order == 0:
+        if x in coeff.free_symbols:
+            return False
+        return True
+    if coeff.is_Mul:
+        if coeff.has(f(x)):
+            return False
+        return x**order in coeff.args
+    elif coeff.is_Pow:
+        return coeff.as_base_exp() == (x, order)
+    elif order == 1:
+        return x == coeff
+    return False
+
+
+def _get_euler_characteristic_eq_sols(eq, func, match_obj):
+    r"""
+    It returns the solution of homogeneous part of the linear euler ODE and
+    the list of roots of characteristic equation.
```

Comment:
Don't use "It" here. Just "Returns ..." is fine.

This doesn't really explain the function very well. What is the `match_obj` parameter supposed to be for example?